### DISCUSSION TITLE
Xacro like functionality in mujoco's xml

### DISCUSSION THREAD

**Discussion by pab47** [google-deepmind/mujoco#106] [General]
Is there a way to define variables in XML and use them, much like how xacro's are used in urdf? For example, in urdf on can define the variable radius and then assign it to say 4.3. Something like this in urdf.
<xacro:property name="radius" value="4.3" /> <circle diameter="${2 * radius}" />

**Comment by saran-t** [comment]
MJCF doesn't support macros, however it does provide cascading `<default>` mechanism. This allows you to define classes of objects that share the same property (e.g. radius in your example). Please refer to the [Defaults settings section](https://mujoco.readthedocs.io/en/latest/modeling.html#cdefault) in our documentation for further detail.

**Comment by brehman-hiab** [comment]
Hello @saran-t 
I am also looking for something similar, I tried using `default` but I am not able to understand how to do it correctly. 
My objective is to spawn multiply instances of the robot given a different origins position, and also maintain only one or two files instead of writing one big XML.

For example, I wrote two XML files separately for two robots:

(file 1)-Mobile wheel robot (the free joint that can take a parameter to modify origin pos) 
(file 2)-fixed based robotic arm (this file can take two parameters as an input to modify the default origin pos of the base link and parent body)

(file 3) then write another XML file that can use these two robot models
include mobile robot file
include  based robotic arm file give pos as the new origin and parent body name

furthermore, write another file that can insert file3 to spawn multiply robot given a different origin

I know, name of the body, joints and actuator cannot be repeated for that I want to append name for example 
`body name="${name}_base"`

I have written the first two files just for discussion 
(file 1)
```
<mujoco model="mobilebase">
    <include file="baisc_scene.xml"/> 
    <compiler inertiafromgeom="true" angle ="degree"/> <!-- telling the we will use degree-->
    <option timestep ="0.001"/> <!-- Setting sample time as 1ms -->
    <worldbody>
        <body name="mobile_torso" pos="0 0 0.65" euler="0 0 0">
            <geom type="box" size="1 0.5 0.25 " rgba=".8 .2 .1 1"/>
            <freejoint name="root"/>
            <body name="right_FW" pos="0.5 -0.6 -0.1">
                <geom type="cylinder" size="0.25 0.1" rgba=".8 .2 .1 1" euler="90 0 0"/>
            </body>
            <body name="left_FW" pos="0.5 0.6 -0.1">
                <geom type="cylinder" size="0.25 0.1" rgba=".8 .2 .1 1" euler="90 0 0"/>
            </body>
            <body name="right_RW" pos="-0.5 -0.6 -0.1">
                <geom type="cylinder" size="0.25 0.1" rgba=".8 .2 .1 1" euler="90 0 0"/>
            </body>
            <body name="left_RW" pos="-0.5 0.6 -0.1">
                <geom type="cylinder" size="0.25 0.1" rgba=".8 .2 .1 1" euler="90 0 0"/>
            </body>
        </body>
    </worldbody>
</mujoco>

```
(file 2)
```
<mujoco model="robotarm">
    <include file="baisc_scene.xml"/> <!-- I am loading defult scene after removing a body that I dont need and can be orride as we want, file exist in same folder-->
    <compiler inertiafromgeom="true" angle ="degree"/> <!-- telling the we will use degree-->
    <option timestep ="0.001"/> <!-- Setting sample time as 1ms -->
    <worldbody>
        <body name="base" pos="0 0 0.0" euler="0 0 0">
            <geom type="cylinder" size="0.25 0.01" rgba=".8 .2 .1 1"/>           
            <body name="column" pos="0.0 0.0 0.0" euler="0 0 0">
                <geom type="capsule" size="0.035 0.25" pos="0 0 0.3" euler="0 0 0"/>
                
                    <body name="link1" pos="0.0 0.0 0.6" euler="90 0 0">
                        <geom type="capsule" size="0.035 0.25" pos="0 0 0.3" euler="0 0 0"/>
                        <joint name="pivot1" type="hinge" pos="0 0 0" axis="1 0 0" limited="false" damping="0.05"></joint>
                        
                            <body name="link2" pos="0.0 0.0 0.6" euler="0 0 0">
                                <geom type="capsule" size="0.035 0.25" pos="0 0 0.3" euler="0 0 0"/>
                                <joint name="pivot2" type="hinge" pos="0 0 0" axis="1 0 0" limited="false" damping="0.05"></joint>
                            </body>
                    </body>
            </body>
        </body>
    </worldbody>
        <actuator>
            <motor joint="pivot1" name="motor1" gear="1" forcelimited="true" forcerange="-100 100"/> <!--joint="pivot" is defined as hinge type in world body -->
            <motor joint="pivot2" name="motor" gear="1" forcelimited="true" forcerange="-100 100"/> <!--joint="pivot" is defined as hinge type in world body -->
        </actuator>
</mujoco>
```

**Comment by yuvaltassa** [comment]
MuJoCo does not currently support model composition, other than the [`<include>` clause](https://mujoco.readthedocs.io/en/latest/modeling.html#including-files) which may or may not meet your needs.

While we are working to improve native model composition, your best bet is `dm_control`'s [`mjcf` module](https://github.com/deepmind/dm_control/blob/master/dm_control/mjcf/README.md).
